src/util/locking/RedisResourceLocker.ts

```diff
@@ -0,0 +1,169 @@
+import { assert } from 'console';
+import type { RedisClient } from 'redis';
+import { createClient } from 'redis';
+import type { Lock } from 'redlock';
+import Redlock from 'redlock';
+import type { ResourceIdentifier } from '../../ldp/representation/ResourceIdentifier';
+import { getLoggerFor } from '../../logging/LogUtil';
+import { InternalServerError } from '../errors/InternalServerError';
+import type { ResourceLocker } from './ResourceLocker';
+
+// The ttl set on a lock, not really important cause redlock wil not handle expiration
+const ttl = 10000;
+// The default redlock config
+const defaultRedlockConfig = {
+  // The expected clock drift; for more details
+  // see http://redis.io/topics/distlock
+  // Multiplied by lock ttl to determine drift time
+  driftFactor: 0.01,
+  // The max number of times Redlock will attempt
+  // to lock a resource before erroring
+  retryCount: 1000000,
+  // The time in ms between attempts
+  retryDelay: 200,
+  // The max time in ms randomly added to retries
+  // to improve performance under high contention
+  // see https://www.awsarchitectureblog.com/2015/03/backoff.html
+  retryJitter: 200,
+};
+/**
+ * A locking system that uses a Redis server or any number of
+ * Redis nodes / clusters
+ * This solution has issues though:
+ *  - Redlock wants to handle expiration itself, this is against the design of a ResourceLocker.
+ *    The workaround for this is to extend an active lock indefinitely.
+ *  - This solution is not multithreaded!
+ *  - Redlock does not provide the ability to see which locks have expired
+ */
+export class RedisResourceLocker implements ResourceLocker {
+  protected readonly logger = getLoggerFor(this);
+
+  public redlock: Redlock;
+  private readonly lockList: Map<string, Lock>;
+  private readonly intervals: Map<string, NodeJS.Timeout>;
+
+  public constructor(redisClients: string[], redlockOptions?: Record<string, number>) {
+    this.lockList = new Map();
+    this.intervals = new Map();
```

Comment:
This would make the code way more complex imo.

```
this.intervals.delete(identifier);
```

Would become

```
this.helperMap.set(identifier, { lock: this.helperMap.get(identifier)?.lock, interval: undefined });
```


We would also be saving a lot of `undefined` in that Map. Trying to `map.delete(identifier)` would need us to first check if there is still an interval present and also clearing it if so, before being able to safely delete that entry.
I am also not sure why its beneficial to try and keep 2 completely independent objects in sync. Both maps are there to overcome a different limitation in the design of Redlock or the acquire-release pattern.

Reply:
This is not a major issue so it can stay like it is if it would be a lot of work to change, I just think it would make it easier programming wise to make sure everything stays consistent and cleaned up correctly.

How I see it: if you acquire a lock, 2 things happen: a lock object gets made so we can release it, and a timer gets made to prevent the timeout of that lock. On release, both that lock and timer are deleted. So in my opinion they're not really independent. I would expect the `acquire` function to create the lock and timer and put them in the map, and the `release` function to clear the lock and timer, and remove them from the map.

This is almost already what happens in the code above:
```
this.lockList.set(resource, lock);
this.extendLockIndefinitely(resource);
```
Instead of this I would then expect something like
```
const timeout = this.extendLockIndefinitely(resource);
this.lockMap.set(resource, { lock, timeout };
```
And similar for the release.

How would there be a lot of `undefined` being saved since every lock needs a timeout right, or am I missing something there?